read OPENSSLDIR from registry (#24450)
openssl/openssl: NOTES-WINDOWS.md

Installation directories
------------------------

On most Unix platform installation directories are determined at build time via

> constant defines.  On Windows platforms however, installation directories are
> determined via registry keys, as it is common practice to build OpenSSL and
> install it to a variety of locations.

Reviewer: ```suggestion
constant defines.  On Windows platforms however, it is also possible to
determine the installation directories via registry keys, as it is common
practice to build OpenSSL and install it to a variety of locations.
```

Author: I'm not sure about this, given that we no longer use the constant defines on windows (per previous conversation), I don't think the phrase "It is also" applies here.  With this change set, you either define OSSL_WINCTX and set the registry keys, set the OPENSSL_CONF, etc environment variables, or it just doesn't try to load anything.  I can add language about the environment variables here if you like

Reviewer: The behaviour should NOT be changed from what we currently have if OSSL_WINCTX is not defined. That means these will be constant defines even on Windows. And AFAIK that is what the current code does.

The following keys:

    `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL-<version>-<ctx>\OPENSSLDIR`
    `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL-<version>-<ctx>\ENGINESDIR`
    `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL-<version>-<ctx>\MODULESDIR`

Can be administratively set, and openssl will take the paths found there as the
values for OPENSSLDIR, ENGINESDIR and MODULESDIR respectively.

To enable the reading of registry keys from windows builds, add
`-DOPENSSL_WINCTX=<string>`to the Configure command line.  This define is used
at build-time to construct library build specific registry key paths of the
format:
`\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432node\OpenSSL-<version>-<ctx>`

> Where `<version>` is the semantic major.minor.patch version of the library being

Reviewer 2: Drop the word "semantic" here. We strictly speaking don't do semantic versioning (just something that looks a lot like it).

built, and `<ctx>` is the value specified by `-DOPENSSL_WINCTX`.  This allows
for multiple openssl builds to be created and installed on a single system, in
which each library can use its own set of registry keys.

Note the installer available at <https://github.com/openssl/installer> will set
these keys when the installer is run.

If the registry keys above do not exist on a given system, or if the
`OSSL_WINCTX` variable is not defined at build time, OpenSSL makes no attempt to

> load configuration, engines of modules from disk.

Reviewer 2: engines **or** modules?

Author: replaced this with a table,

> load configuration, engines of modules from disk.

Reviewer 2: Hmm...this text doesn't seem to quite marry with what I see in the implementation. It seems to me that if `OSSL_WINCTX` is not defined that we fall back to build time OPENSSLDIR etc??

Author: yeah, miswrote that, its fixed as a table momentarily

> If the registry keys above do not exist on a given system, or if the
> `OSSL_WINCTX` variable is not defined at build time, OpenSSL makes no attempt to
> load configuration, engines of modules from disk.

Reviewer: This is not true after the recent fixup.  IMO we should not break the existing builds for which
OSSL_WINCTX is not defined and that is what is implemented in defaults.c currently.

```suggestion
If the registry keys above do not exist on a given system and the `OSSL_WINCTX`
variable was defined at build time, OpenSSL makes no attempt to load
configuration, engines of modules from disk.
```

Author: This might be better represented as a table

Installation directories
------------------------

On most Unix platform installation directories are determined at build time via
constant defines.  On Windows platforms however, installation directories are
determined via registry keys, as it is common practice to build OpenSSL and
install it to a variety of locations.

The following keys:

    `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL-<version>-<ctx>\OPENSSLDIR`
    `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL-<version>-<ctx>\ENGINESDIR`
    `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL-<version>-<ctx>\MODULESDIR`

Can be administratively set, and openssl will take the paths found there as the
values for OPENSSLDIR, ENGINESDIR and MODULESDIR respectively.

To enable the reading of registry keys from windows builds, add
`-DOPENSSL_WINCTX=<string>`to the Configure command line.  This define is used
at build-time to construct library build specific registry key paths of the
format:
`\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432node\OpenSSL-<version>-<ctx>`

> Where `<version>` is the major.minor.patch version of the library being

Author: Its a different registry key for every major.minor.patch version, the location can be set to the same path for each key

> Where `<version>` is the major.minor.patch version of the library being

Reviewer 2: My point is it shouldn't be a different key for different patch versions but the same major.minor

is, of course, to choose a different set of directories by using
`--prefix` and `--openssldir` when configuring.

Note that, on Windows platforms (both 32 and 64 bit), the above build time
defaults can be overridden by registry keys.  This is done because it is common
practice for windows based installers to allow users to place the installation

> tree at various locations not defined at run time.  The following keys:

Reviewer 2: Did you mean "not defined at build time"?

    `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL\OPENSSLDIR`
    `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL\ENGINESDIR`
    `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL\MODULESDIR`

Can be administratively set, and openssl will take the paths found there as the
values for OPENSSLDIR, ENGINESDIR and MODULESDIR respectively.  If unset, the
build time defaults will be used.

Note the installer available at https://github.com/openssl/installer will set

> theys keys when the installer is run.

Reviewer 2: typo: these

is, of course, to choose a different set of directories by using

> `--prefix` and `--openssldir` when configuring.

Author: I should modify the NOTE below as well to not reflect the fact that the build time defaults are no longer used

> Note that, on Windows platforms (both 32 and 64 bit), the above build time

Reviewer 3: ```suggestion
Note that, on Windows platforms (both 32 and 64 bit), the above build-time
```

defaults can be overridden by registry keys.  This is done because it is common

> practice for windows based installers to allow users to place the installation

Reviewer 3: ```suggestion
practice for windows-based installers to allow users to place the installation
```

> tree at various locations not defined at build time.  The following keys:

Reviewer 3: ```suggestion
tree at an arbitrary location that differs from the build-time value.  The following keys:
```

    `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL-<version>-<ctx>\OPENSSLDIR`
    `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL-<version>-<ctx>\ENGINESDIR`

>     `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSSL-<version>-<ctx>\MODULESDIR`

Author: yes, reg.exe inserts the Wow6432Node element into the path when /reg:32 or /reg:64 is specified.

However, when looking at the key paths with regedit, the node is explicitly displayed, so I thought it would be better to explicitly include it here so it visually matched up to an adminstrator.

Can be administratively set, and openssl will take the paths found there as the
values for OPENSSLDIR, ENGINESDIR and MODULESDIR respectively.

To enable the reading of registry keys from windows builds, add
`-DOPENSSL_WINCTX=<string>`to the Configure command line.  This define is used
at build time to construct library build specific registry key paths of the
format:

> `\\HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432node\OpenSSL-<version>-<ctx>`

Author: as above, its wierd, I know.  Windows....
